src/server.ts

```diff
@@ -346,37 +336,41 @@ wss.on('connection', async (ws: Server.Ws) => {
           activeSubscriptions[clientId].splice(activeAddressSubIndex);
         }
 
-        const message = prepareMessage(data.id, clientId, {
-          subscribed: false,
-        });
+        response = prepareMessage({ id, clientId, data: { subscribed: false } });
 
-        ws.send(message);
         addressesSubscribed[clientId] = [];
 
         break;
       }
 
       case MESSAGES.PUSH_TRANSACTION: {
-        const submitTransactionMessage = await submitTransaction(
-          data.id,
-          clientId,
-          data.params.txData,
-        );
+        assertParams(params, { txData: 'string' });
+        response = await submitTransaction(id, clientId, params.txData);
 
-        ws.send(submitTransactionMessage);
         break;
       }
 
       default: {
-        const message = prepareErrorMessage(
-          data.id,
-          `Unknown message id: ${data.command}`,
-          clientId,
-        );
-
-        ws.send(message);
+        response = prepareErrorMessage(id, clientId, `Unknown command: ${command}`);
       }
     }
+
+    ws.send(response);
+  };
+
+  ws.on('message', (message: string) => {
+    const data = { id: -1 } as Messages;
```

Comment:
To better understand this we need to take a look at the first line of `onMessage`, the function called immediately after this line.

The first line of `onMessage` is:
```
 onMessage = async (message: string, data: Messages) => {
    Object.assign(data, getMessage(message));
```

which
- in case of success, injects the parse message in `data` (overwriting the `id` with the one got from the message);
- in case of error, throws leaving untouched the `data.id` value (`-1`).

This simplify the `prepareErrorMessage` some lines later:
in that call `data.id` is initialized to `-1` in case the error happens before the message is parsed by `getMessage` (i.e. no `id` is yet available), or the one got from the message if the error happens later.

Without this _dummy object_ we'd need to add a specific error handling for errors thrown inside `getMessage`.

In my opinion the simplification the code have from the presence of this _dummy object_ is a good compromise, if you don agree, I can refactor this.

Reply:
I do understand how this code works, I just find it way too obscure for what it is. Wouldn't be something like this more straightforward?

```ts
 ws.on('message', (message: string) => {
    let parsedMessage: Messages;

    // Attempt to parse the received ws message.
    // If parsing fails, send an error response back to the client and exit early.
    try {
      parsedMessage = getMessage(message);
    } catch (error) {
      const msgError = prepareErrorMessage(
        -1,
        clientId,
        error instanceof MessageError ? error.reason : error,
      );

      return ws.send(msgError);
    }

    // Process the parsed message
    onMessage(parsedMessage).catch(error => {
      logger.error(error);
      const response = prepareErrorMessage(
        parsedMessage.id,
        clientId,
        error instanceof MessageError ? error.reason : error,
      );

      ws.send(response);
    });
  });
```

```ts
 const onMessage = async (data: Messages) => {
    const { command, id, params } = data;
    let response: string;
...

```

I did not have to override type checking and I don't have to mutate msg object inside the onMessage function.

Reply:
Ok, let me try to fix this as well

Reply:
551bada should solve this

Reply:
nice, thats looks much better! 👍